fid.1
.TH FID 1
.SH NAME
fid \- query id database for specific files
.SH SYNOPSIS
.B fid
.RB [ \-f \^file]
file1 [ file2 ]
.SH DESCRIPTION
.I Fid
is a query tool for the id database. If you specify a single file
name as an argument, it prints a list of all the identifiers that
occur in that file.
.PP
When you give it two file names it takes the intersection. It prints
only the list of identifiers that occur in both files.
.PP
The following options are recognized:
.TP 10
.BR \-f file\^
Use
.I file\^
as the database instead of the default
.BR ID .
.SH SEE ALSO
mkid(1),
lid(1).


NEWS
id-utils NEWS - User visible changes.

Version 3.2e

* FIXME

Version 3.2d - April 6, 1999, by Greg McGary

* Modernize lisp/Makefile.am, replace gets with fgets.

* Allocate larger scanner buffer: MIN (2MB, largest file)

* compile lib/regex.c if regcomp is absent in libc.

Version 3.2c - April 3, 1999, by Greg McGary

* Add verbose traces to file walker: `mkid -V'

* New heuristic for computing initial size of token table:
MAX (1M, lg(n) * n * 16)

Version 3.2b - March 29, 1999, by Greg McGary

* Fixed assorted bugs

* Remove arbitrary limit of 32767 on number of files that
  can be scanned.  Now the number of files is only
  limited by available virtual memory.

Version 3.2a - November 22, 1998, by Greg McGary

* Fixed assorted bugs

Version 3.2 - August 6, 1996, by Greg McGary

* Major bug in eid fixed

Version 3.1 - July 17, 1996, by Greg McGary

* A few more portability bugs are fixed

Version 3.0m - July 9, 1996, by Greg McGary

* NLS is now initialized properly

* Assorted portability problems are fixed

* A few bugs (one serious) are fixed

Version 3.0l - skipped `l' (ell) is too easy to confuse with `1' (one)

Version 3.0k - June 5, 1996, by Greg McGary

* `lid's command line has been revised and expanded.  Control over output
  formats is expanded.  Pattern searching options are now complete and
  orthogonal: word/substring, literal/regexp, case (in)sensitivity.

* The texinfo manual, id-utils.texi, is now up-to-date.

* `mkid' and `xtokid' support the new `--prune' option.

* `mkid' and `xtokid' now implement the `--include' and `--exclude 'options.

* `lid -R grep' (a.k.a. `gid') no longer prepends a space to each line--
  now its output is exactly like grep's.

* `fid' separates tokens with spaces if output is to a termianal, and
  otherwise with newlines.

* If lid is in `brace' mode by default, it uses brace notation when
  outputting to a terminal and separates file names with spaces otherwise.
  If printing the token or pattern key is disabled, then file names are
  separated with newlines.

* All programs now accept directory names for the `-f, --file=' args.
  When a directory is given, the file name "ID" is assumed.

* gid.el is renamed id-utils.el and is installed in emacs's site-lisp/

* Source files and directories are reorganized

* Misc. bugs are fixed

Version 3.0j - May 31, 1996, by Greg McGary

* Pre-test versions are now designated with a lower-case letter to shorten
  the top-level directory name to 14 chars.

* `idx' has been renamed `xtokid'.

* The old `pid' program has been reborn as `fnid'.

* Some bugs & build problems are fixed.

* Some missing prototypes are added to header files.

* gcc -Wall (plus a bunch of other other warnings) is happier.

* A number of idtest problems are fixed.

Version 3.0.9 - May 30, 1996, by Greg McGary

* The package name has been changed from `mkid' to `id-utils', and is now
  officially a part of GNU.

* Command-line arguments have been revised to conform with GNU standards.
  Each command now processes --help and --version long arguments.

* The texinfo manual is out-of-date.  Please use the `--help' option for
  correct information about command-line arguments.

* Message catalogs for French and German exist, but are out-of-date.

* `mkid' and `idx' use a file-tree walker.  Give them one or more directory
  names on the command-line, and they'll recursively descend looking for
  source files.  lib/language.map (installe as $(prefix)/share/language.map)
  contains patterns for mapping file names to the appropriate language scanner.

* id-utils now uses automake.

* gid, aid and eid are now shell-script wrappers that
  call `lid' with the apropriate argument.

* A new environment `IDPATH' contains a colon-separated list of ID file
  names for use by `fid' and `lid'.  It is only partially implemented at
  this time only the first ID file is actually handled, the rest are
  ignored.

* pid is temporarily gone, but will reappear soon.

* iid has been removed from the distribution.

* Support for non-ANSI compilers is broken.  Don't expect it to work.
  If you use gcc, all will be well.

* Miscellaneous bugs are fixed.

Version 3.0 - February 1995, by Greg McGary

* `mkid' uses a new efficient algorithm for building the database.  The old
  algorithm was O(n^2) for space and would exhaust memory or lead to page
  thrashing on very large programs.  The new algorithm is linear with the
  size of the input for both time and space.  It is now possible to build
  very large `mkid' databases covering many thousands of files totalling
  tens of megabytes and containing hundreds of thousands of unique tokens.

* The database header is now byte-order and word-size independent.  ID
  files may be shared across networks of dissimilar architectures.

Version 2.0 - March 1991, by Tom Horsley


Version 1.0 - September 1987, by Greg McGary

* Initial release to comp.sources.unix

Local Variables:
mode: text
fill-column: 75
End:


src/fid.c
/* fid.c -- list all tokens in the given file(s)
   Copyright (C) 1986, 1995, 1996 Free Software Foundation, Inc.
   Written by Greg McGary <gkm@gnu.ai.mit.edu>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. */

#include <config.h>
#include <stdio.h>
#include <getopt.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <errno.h>
#include <alloca.h>
#include <error.h>
#include <pathmax.h>
#include <xalloc.h>

#include "xnls.h"
#include "idfile.h"
#include "iduglobal.h"

int get_file_index (char *file_name);
int is_hit (unsigned char const *hits, int file_number);
int is_hit_1 (unsigned char const **hits, int level, int file_number);
void skip_hits (unsigned char const **hits, int level);
void usage (void);

struct idhead idh;
int tree8_levels;

/* The name this program was run with. */

char const *program_name;

/* If nonzero, display usage information and exit.  */

static int show_help;

/* If nonzero, print the version on standard output then exit.  */

static int show_version;

/* The file name of the ID database.  */

struct file_link *cw_dlink;
struct file_link **members_0;
unsigned int bits_vec_size;
char *hits_buf;

static struct option const long_options[] =
{
  { "file", required_argument, 0, 'f' },
  { "help", no_argument, &show_help, 1 },
  { "version", no_argument, &show_version, 1 },
  { 0 }
};

void
usage (void)
{
  fprintf (stderr, _("Try `%s --help' for more information.\n"),
	   program_name);
  exit (1);
}

static void
help_me (void)
{
  printf (_("\
Usage: %s [OPTION] FILENAME [FILENAME2]\n\
"), program_name);
  printf (_("\
List identifiers that occur in FILENAME, or if FILENAME2 is\n\
also given list the identifiers that occur in both files.\n\
\n\
  -f, --file=FILE  file name of ID database\n\
      --help       display this help and exit\n\
      --version    output version information and exit\n\
"));
  printf (_("\nReport bugs to " PACKAGE_BUGREPORT "\n\n"));
  exit (0);
}

int
main (int argc, char **argv)
{
  int index_1 = -1;
  int index_2 = -1;

  program_name = argv[0];
  idh.idh_file_name = 0;
  

#if ENABLE_NLS
  /* Set locale according to user's wishes.  */
  setlocale (LC_ALL, "");

  /* Tell program which translations to use and where to find.  */
  bindtextdomain (PACKAGE, LOCALEDIR);
  textdomain (PACKAGE);
#endif

  for (;;)
    {
      int optc = getopt_long (argc, argv, "f:",
			      long_options, (int *) 0);
      if (optc < 0)
	break;
      switch (optc)
	{
	case 0:
	  break;

	case 'f':
	  idh.idh_file_name = optarg;
	  break;

	default:
	  usage ();
	}
    }

  if (show_version)
    {
      printf ("%s - %s\n", program_name, PACKAGE_VERSION);
      exit (0);
    }

  if (show_help)
    help_me ();

  argc -= optind;
  argv += optind;
  if (argc < 1)
    {
      error (0, 0, _("no file name arguments"));
      usage ();
    }
  if (argc > 2)
    {
      error (0, 0, _("too many file name arguments"));
      usage ();
    }

  /* Look for the ID database up the tree */
  idh.idh_file_name = locate_id_file_name (idh.idh_file_name);
  if (idh.idh_file_name == 0)
    error (1, errno, _("can't locate `ID'"));

  init_idh_obstacks (&idh);
  init_idh_tables (&idh);

  cw_dlink = get_current_dir_link ();

  /* Determine absolute name of the directory name to which database
     constituent files are relative. */
  members_0 = read_id_file (idh.idh_file_name, &idh);
  bits_vec_size = (idh.idh_files + 7) / 4; /* more than enough */
  tree8_levels = tree8_count_levels (idh.idh_files);

  index_1 = get_file_index ((argc--, *argv++));
  if (argc)
    index_2 = get_file_index ((argc--, *argv++));

  if (index_1 < 0)
    return 1;

  hits_buf = xmalloc (idh.idh_buf_size);
  fseek (idh.idh_FILE, idh.idh_tokens_offset, SEEK_SET);
  {
    int count = 0;
    int i;
    int separator = (isatty (STDOUT_FILENO) ? ' ' : '\n');
    
    for (i = 0; i < idh.idh_tokens; i++)
      {
	unsigned char const *hits;

	gets_past_00 (hits_buf, idh.idh_FILE);
	hits = token_hits_addr (hits_buf);
	if (is_hit (hits, index_1) && (index_2 < 0 || is_hit (hits, index_2)))
	  {
	    fputs (token_string (hits_buf), stdout);
	    putchar (separator);
	    count++;
	  }
      }
    if (count && separator == ' ')
      putchar ('\n');
  }

  return 0;
}

int
get_file_index (char *file_name)
{
  struct file_link **members;
  struct file_link **end = &members_0[idh.idh_files];
  struct file_link *fn_flink = 0;
  char *file_name_buf = alloca (PATH_MAX);
  int has_slash = (strchr (file_name, '/') != 0);
  int file_name_length = strlen (file_name);
  int idx = -1;

  if (strstr (file_name, "./"))
    fn_flink = parse_file_name (file_name, cw_dlink);

  for (members = members_0; members < end; members++)
    {
      struct file_link *flink = *members;
      if (fn_flink)
	{
	  if (fn_flink != flink)
	    continue;
	}
      else if (has_slash)
	{
	  int member_length;
	  maybe_relative_file_name (file_name_buf, flink, cw_dlink);
	  member_length = strlen (file_name_buf);
	  if (file_name_length > member_length
	      || !strequ (&file_name_buf[member_length - file_name_length], file_name))
	    continue;
	}
      else if (!strequ (flink->fl_name, file_name))
	continue;
      if (idx >= 0)
	{
	  error (0, 0, _("`%s' is ambiguous"), file_name);
	  return -1;
	}
      idx = members - members_0;
    }
  if (idx < 0)
    error (0, 0, _("`%s' not found"), file_name);
  return idx;
}

int
is_hit (unsigned char const *hits, int file_number)
{
  return is_hit_1 (&hits, tree8_levels, file_number);
}

int
is_hit_1 (unsigned char const **hits, int level, int file_number)
{
  int file_hit = 1 << ((file_number >> (3 * --level)) & 7);
  int hit = *(*hits)++;
  int bit;

  if (!(file_hit & hit))
    return 0;
  if (level == 0)
    return 1;

  for (bit = 1; (bit < file_hit) && (bit & 0xff); bit <<= 1)
    {
      if (hit & bit)
	skip_hits (hits, level);
    }
  return is_hit_1 (hits, level, file_number);
}

void
skip_hits (unsigned char const **hits, int level)
{
  int hit = *(*hits)++;
  int bit;

  if (--level == 0)
    return;
  for (bit = 1; bit & 0xff; bit <<= 1)
    {
      if (hit & bit)
	skip_hits (hits, level);
    }
}
